gammapy/utils/random/inverse_cdf.py

```diff
@@ -77,3 +83,125 @@ def sample(self, size):
+    ):
+        self.random_state = get_random_state(random_state)
+        self.npred_map = npred_map
+        self.lc = lc
+        self.phase_lc = phase_lc
+        self.tmin = tmin
+        self.tmax = tmax
+
+    def npred_total(self):
+        """ Calculate the total number of the sampled predicted events.
+            
+        Returns
+        -------
+        random_state.poisson() : int
+                        Number of predicted events.
+        """
+
+        return self.random_state.poisson(np.sum(self.npred_map.data))
+
+    def sample_npred(self):
+        """ Calculate energy and Galactic coordinates of the sampled source events.
+            
+        Returns
+        -------
+        coords : `~numpy.ndarray`
+                Array with coordinates and energies of the sampled events.
+        """
+
+        self.n_events = self.npred_total()
+
+        cdf_sampler = InverseCDFSampler(
+            self.npred_map.data, random_state=self.random_state
+        )
+
+        pix_coords = cdf_sampler.sample(self.n_events)
+        coords = self.npred_map.geom.pix_to_coord(pix_coords[::-1])
+
+        return coords
+
+    def sample_timepred(self):
+        """ Calculate the times of arrival of the sampled source events.
+
+        Returns
+        -------
+        ToA : `~numpy.array`
+            array with times of the sampled events.
+        """
+
+        n_events = self.n_events
+        if (self.lc is not None) and (self.tmax > 0):
+            n_tbin = self.tmax - self.tmin
+            t = np.linspace(self.tmin, self.tmax, n_tbin)
+            normalization = self.lc._interpolator(t, ext=3)
+            time_sampler = InverseCDFSampler(
+                normalization, random_state=self.random_state
+            )
+            ToA = time_sampler.sample(n_events)[0]
+
+        if (self.phase_lc is not None) and (self.tmax > 0):
+            n_tbin = self.tmax - self.tmin
+            t = np.linspace(self.tmin, self.tmax, n_tbin)
+            normalization = self.phase_lc.evaluate_norm_at_time(t)
+            time_sampler = InverseCDFSampler(
+                normalization, random_state=self.random_state
+            )
+            ToA = time_sampler.sample(n_events)[0]
+
+        else:
+            ToA = self.tmin + self.random_state.uniform(
+                high=(self.tmax - self.tmin), size=n_events
+            )
+
+        return ToA
+
+    def sample_events(self):
+        """It converts the given sampled event list into an astropy table.
+            
+        Returns
+        -------
+        events : `~astropy.table`
+            Sampled event list in an astropy table format.
+        """
+
+        coords = self.sample_npred()
+
+        events = Table()
+        events["lon_true"] = coords[0] * u.deg
+        events["lat_true"] = coords[1] * u.deg
+        events["e_true"] = coords[2] * u.TeV
+        if ((self.lc is not None) and (self.tmax > 0)) or (self.tmax > 0):
```

Comment:
Once `t_min` and `t_max` are required arguments this becomes non-optional...

Reply:
Done.